contracts/voucher/Voucher.sol

```diff
@@ -233,6 +226,33 @@ contract Voucher is IVoucher, Upgradeable, ERC721Upgradeable, Utils {
         emit UseGlobalURIUpdated(newUseGlobalURI);
     }
 
+    /**
+     * @dev sets the controller address
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of this contract
+     * - controller address must not be set
+     */
+    function setController(address controllerAddress) public onlyAdmin {
+        if (_controllerSet) {
```

Comment:
A few notes:
- You don't check that `controllerAddress` isn't `0x0` (+ see the comment above about not needing `_controllerSet`)
- `setController` seems a bit misplaced. I'd expect, in a contract like this:
    - For new deployments, the controller is passed a parameter to `initialize` (and this way can be only set once as well).
    - For the current instance, the controller is set via `postUpgrade` (passed externally). If you had kept the `ROLE_MINTER` role, then `postUpgrade` could've had enumerated it and assigned the controller to be the first address with this rule (which is possible since it uses `EnumerableSet`).
- The function can be `external`

Reply:
changed 1 and 3, for 2) it's not necessary, since the upgradeability / proxies are going to be removed